frontend/src/providers/QueryBuilder.tsx

```diff
@@ -64,37 +61,104 @@ export function QueryBuilderProvider({
 		setQueryBuilderData(initialQueryBuilderData);
 	}, []);
 
-	// ** Method for setupping query builder data
+	// ** Method for setuping query builder data
+	// ** Before setuping transform data from backend to frontend format
 	const initQueryBuilderData = useCallback(
 		(queryBuilderData: QueryBuilderData): void => {
 			setQueryBuilderData(queryBuilderData);
 		},
 		[],
 	);
 
+	const removeEntityByIndex = useCallback(
+		(type: keyof QueryBuilderData, index: number) => {
+			setQueryBuilderData((prevState) => {
+				const currentArray: (IBuilderQueryForm | IBuilderFormula)[] =
+					prevState[type];
+				return {
+					...prevState,
+					[type]: currentArray.filter((item, i) => index !== i),
+				};
+			});
+		},
+		[],
+	);
+
+	const addNewQuery = useCallback(() => {
+		setQueryBuilderData((prevState) => {
+			if (prevState.queryData.length >= MAX_QUERIES) return prevState;
+
+			const existNames = prevState.queryData.map((item) => item.queryName);
+
+			const newQuery: IBuilderQueryForm = {
+				...initialQueryBuilderFormValues,
+				queryName: createNewQueryName(existNames),
+				...(initialDataSource
+					? {
+							dataSource: initialDataSource,
+							aggregateOperator: mapOfOperators[initialDataSource][0],
+							expression: createNewQueryName(existNames),
+					  }
+					: {}),
+			};
+
+			return { ...prevState, queryData: [...prevState.queryData, newQuery] };
+		});
+	}, [initialDataSource]);
+
+	const setupInitialDataSource = useCallback(
+		(newInitialDataSource: DataSource | null) =>
+			setInitialDataSource(newInitialDataSource),
+		[],
+	);
+
 	const handleSetQueryData = useCallback(
 		(index: number, newQueryData: Partial<IBuilderQueryForm>): void => {
-			const updatedQueryBuilderData = queryBuilderData.queryData.map((item, idx) =>
-				index === idx ? { ...item, ...newQueryData } : item,
-			);
-
-			setQueryBuilderData((prevState) => ({
-				...prevState,
-				queryData: updatedQueryBuilderData,
-			}));
+			setQueryBuilderData((prevState) => {
```

Comment:
here also

Reply:
This is setter from useState. Do you suggest separate logic inside where prevState to another function?

Reply:
yeah broken down into smaller chunks